[18.0][ADD] sale_product_catalog_extended: New module to use catalog as a product_picker

CarlosRoca13 opened the pull request:
cc @Tecnativa TT52109

ping @carlosdauden @sergio-teruel

CarlosRoca13 committed:
[ADD] sale_product_catalog_extended: New module to use catalog as a product_picker

CarlosRoca13 committed:
[ADD] sale_product_catalog_extended: New module to use catalog as a product_picker

pedrobaeza commented:
Reading the description, this should be called `sale_product_catalog_inline_edit`, isn't it? Or do you plan to add more things?

CarlosRoca13 commented:
> Reading the description, this should be called `sale_product_catalog_inline_edit`, isn't it? Or do you plan to add more things?

More things are added, I'm improving the description

pedrobaeza commented:
OK, anyway, can we avoid the suffix `_extended`, as it's very generic?

carlosdauden committed:
[IMP] sale_product_catalog_extended: Add catalog_origin_data option to select products from last sales

carlosdauden committed:
[IMP] sale_product_catalog_extended: Add catalog_origin_data option to select products from last sales

alexey-pelykh commented on sale_product_catalog_extended/controllers/catalog.py:
```diff
@@ -0,0 +1,33 @@
+        """
+        order = request.env["sale.order"].browse(order_id)
+        return order.with_company(order.company_id)._get_catalog_order_line(
+            product_id,
```
**Security concern**: `order_id` is accepted directly from the client. There is no verification that the current user has access to this sale order. A logged-in user could pass any order ID and retrieve line information for orders they should not see. Consider adding `order.check_access_rights('read')` and `order.check_access_rule('read')`, or use `_search` to limit to accessible records.

alexey-pelykh commented on sale_product_catalog_extended/controllers/catalog.py:
```diff
@@ -0,0 +1,33 @@
+        :rtype: dict
+        """
+        order_lines = request.env["sale.order.line"].browse(order_line_ids)
+        return order_lines._get_product_catalog_lines_data(**kwargs)
```
**Security concern**: Same issue as above -- `order_line_ids` are accepted directly from the client without access verification. A user could pass arbitrary sale order line IDs to retrieve data they should not see.

alexey-pelykh commented on sale_product_catalog_extended/models/product_product.py:
```diff
@@ -0,0 +1,81 @@
+            # Then browse the ids to respect the order.
+            product_ids = getattr(
+                self,
+                f"_get_product_picker_data_{domain[catalog_orig_data_bool_list.index(True)][2]}",
```
This domain inspection is fragile. `domain` can contain string operators (`'&'`, `'|'`, `'!'`) which will also be iterated, and `"catalog_origin_data" in subdomain` will match if the string appears anywhere in the element (including as a substring). Consider using `odoo.osv.expression` utilities to safely extract leaves by field name.

Also, this bypasses `offset`, `limit`, and `order` parameters entirely when the special domain is detected -- it returns a browsed recordset without respecting pagination. This could cause performance issues with large result sets.

alexey-pelykh commented on sale_product_catalog_extended/models/product_product.py:
```diff
@@ -0,0 +1,81 @@
+        other_sales = (
+            self.env["sale.order"]
+            # .sudo()
+            ._search(
```
Commented-out `.sudo()` -- please either remove or, if sudo access is intentionally needed, uncomment with a justification comment explaining why elevated privileges are required.

alexey-pelykh commented on sale_product_catalog_extended/models/product_product.py:
```diff
@@ -0,0 +1,81 @@
+        limit = int(
+            self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("sale_order_product_picker.product_picker_last_order_limit", "0")
```
The config parameter key `sale_order_product_picker.product_picker_last_order_limit` does not match the module name `sale_product_catalog_extended`. This appears to be a leftover from a different/internal module. Consider renaming to `sale_product_catalog_extended.product_picker_last_order_limit` for consistency.

alexey-pelykh commented on sale_product_catalog_extended/models/product_product.py:
```diff
@@ -0,0 +1,81 @@
+            product_ids = getattr(
+                self,
+                f"_get_product_picker_data_{domain[catalog_orig_data_bool_list.index(True)][2]}",
+            )()
```
**Security risk**: This constructs a method name from a domain value (`domain[...][2]`) using `getattr` + f-string. While currently limited to the `catalog_origin_data` selection field, this pattern allows invoking any method named `_get_product_picker_data_*` on the model. A safer approach would be to use an explicit mapping dict:
```python
method_map = {"sale_order": self._get_product_picker_data_sale_order}
method = method_map.get(value)
```

alexey-pelykh commented on sale_product_catalog_extended/models/sale_order.py:
```diff
@@ -0,0 +1,43 @@
+    def _get_action_add_from_catalog_extra_context(self):
+        # TODO: partner_id or shipping_partner_id?
+        return {
+            **super()._get_action_add_from_catalog_extra_context(),
```
TODO comment suggests the design is not finalized. `partner_id` vs `partner_shipping_id` affects which customer's past orders are shown. This should be resolved before the PR is ready for review.

alexey-pelykh commented on sale_product_catalog_extended/models/sale_order.py:
```diff
@@ -0,0 +1,43 @@
+        if len(self) > 1:
+            res["readOnly"] = (
+                self.order_id._is_readonly()
+                or self.product_id.sale_line_warn == "block"
```
`self.order_id._is_readonly()` and `self.product_id.sale_line_warn` are called on a recordset that can contain multiple lines (the `len(self) > 1` branch). If lines belong to different orders, `self.order_id` is a multi-record recordset, and `_is_readonly()` may raise or behave unexpectedly. Consider iterating or using `self.order_id[:1]._is_readonly()` if all lines are guaranteed to share the same order.

alexey-pelykh commented on sale_product_catalog_extended/static/src/product_catalog/order_line/order_line.xml:
```diff
@@ -0,0 +1,41 @@
+                class="btn btn-light border w-100"
+            >
+                <i class="fa fa-pencil-square-o" />
+                Edit
```
The strings "Edit", "Edit Lines", and "Total quantity: " are hardcoded in English. In Odoo OWL templates, user-facing text should be translatable. Wrap them appropriately so they can be extracted for i18n.

alexey-pelykh reviewed (commented):
## Review: sale_product_catalog_extended

This PR is currently marked as **Draft**. The comments below are intended as early feedback to help guide development.

### 1. Security: Missing access control on controller endpoints

Both new JSON-RPC endpoints (`/product/catalog/sale/open_order_line` and `/product/catalog/sale/get_order_line_data`) accept `order_id` / `order_line_ids` directly from the client but do not perform any access rights verification beyond `auth="user"`. A logged-in user could pass arbitrary IDs to read order line data belonging to other users/companies. Consider:
- Adding explicit `check_access_rights` / `check_access_rule` calls, or
- Using `_search` with the current user's context to filter accessible records.

### 2. No security files (`ir.model.access.csv`)

The module has no `security/` directory and no access rules declared in `__manifest__.py`. Even if the module only inherits existing models, OCA guidelines typically expect an explicit security directory if new fields or controllers are introduced.

### 3. No tests

There is no `tests/` directory. OCA modules require test coverage. At minimum, tests should cover:
- The new controller endpoints (access control, expected return values)
- The `search_fetch` override behavior
- The `_get_product_catalog_lines_data` logic for single and multi-line cases

### 4. Commented-out `.sudo()` in `product_product.py`

```python
# .sudo()
```

This looks like leftover debug code (line ~231 in the diff). It should either be removed or, if sudo is intentionally needed, uncommented with a justification comment.

### 5. Fragile `search_fetch` override

The override inspects domain elements by checking `"catalog_origin_data" in subdomain`, which treats each domain leaf as iterable/string-containable. This is fragile because:
- It will break on non-tuple domain elements (operators like `'&'`, `'|'`, `'!'`).
- It does not handle nested domains or normalized domain formats.

Consider using `odoo.osv.expression` utilities to safely extract the relevant leaf.

### 6. Config parameter key mismatch

The `ir.config_parameter` key is `sale_order_product_picker.product_picker_last_order_limit` but the module is named `sale_product_catalog_extended`. This looks like a leftover from a different/internal module. Should be renamed to match this module's technical name.

### 7. `_get_product_catalog_lines_data` multi-record safety

In `sale_order.py`, `_get_product_catalog_lines_data` accesses `self.order_id` and `self.product_id` directly on the recordset without iteration. When `len(self) > 1`, `self.order_id` returns a recordset of all related orders (or raises if `ensure_one` is expected internally by `_is_readonly`). The `.order_id._is_readonly()` call may not behave correctly for multiple distinct orders.

### 8. Hardcoded English strings in XML templates

The strings "Edit", "Edit Lines", and "Total quantity: " in `order_line/order_line.xml` are not wrapped for translation. In Odoo OWL templates, user-facing strings should use the translation mechanism.

### 9. TODO comments

There are `TODO` comments in:
- `sale_order.py` line ~291: `# TODO: partner_id or shipping_partner_id?`
- `sale_order_line_views.xml`: `<!-- TODO: Change to target new...`

These suggest the implementation is not finalized.

### 10. `getattr` with `f-string` for dynamic dispatch

In `product_product.py`:
```python
getattr(self, f"_get_product_picker_data_{domain[...][2]}")
```

This constructs a method name from user-controllable domain values. While domain values come from the search panel (limited selection), this pattern is risky. If the value is ever extended or manipulated, it could invoke unintended methods. Consider a mapping dict instead.

### Summary

Since this is a draft PR, these are early observations. The main blockers for OCA merge-readiness would be:
- **Security**: Access control on endpoints and the dynamic dispatch pattern
- **Tests**: Required by OCA
- **Code cleanup**: Commented code, TODOs, config key naming

---
*Review posted via [CorporateHub](https://corporatehub.eu) OCA review campaign*

CarlosRoca13 committed:
[IMP] sale_product_catalog_extended: Add picker features

CarlosRoca13 commented:
I think it is ready to be merged @carlosdauden @sergio-teruel :smile:

CarlosRoca13 committed:
[IMP] sale_product_catalog_extended: Add a system parameter to select default origin data

CarlosRoca13 committed:
[IMP] sale_product_catalog_extended: Remove unnecesary action and add numeric step to product_uom_qty

CarlosRoca13 committed:
[IMP] sale_product_catalog_extended: Set image under the three-dots dropdown to avoid collapse

CarlosRoca13 committed:
[IMP] sale_product_catalog_extended: Add new system parameter to preselect price_mode

CarlosRoca13 committed:
[IMP] sale_product_catalog_extended: catalog sale history by commercial partner

CarlosRoca13 committed:
[FIX] sale_product_catalog_extended: Execute onchages when add a product

CarlosRoca13 committed:
[FIX] sale_product_catalog_extended: Allow extra filters with last sales origin

When selecting the "Last sales" origin in the product catalog, search_fetch returned the origin products directly, discarding the rest of the domain. This made any additional filter (category, name, etc.) ineffective.

Now the remaining catalog filters are applied on top of the origin products while keeping the ordering provided by the origin method.

CarlosRoca13 committed:
[FIX] sale_product_catalog_extended: Avoid flash when adding line with last price

CarlosRoca13 committed:
[FIX] sale_product_catalog_extended: filter categories by last sales origin

When using the "Last sales" data origin, the category search panel showed categories of products that were not on screen and omitted others that were, because the category computation did not respect the selected origin.

eduezerouali-tecnativa commented on sale_product_catalog_extended/models/sale_order.py:
```diff
@@ -0,0 +1,202 @@
+        """Return the user/company default value set for ``use_delivery_address``
+        through Default Values (ir.default).
+        """
+        field = self.env["ir.model.fields"].search(
```
```suggestion
        field = self.env["ir.model.fields"].sudo().search(
```
user might not be allowed to access it

eduezerouali-tecnativa reviewed (changes requested):
Tested on runboat.
